### ISSUE TITLE
Preload next track when approaching the end of the current one

### ISSUE THREAD

**Issue by sashahilton00** [librespot-org/issue-test#2]
<a href="https://github.com/plietar"><img src="https://avatars0.githubusercontent.com/u/1489775?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [plietar](https://github.com/plietar)**
_Wednesday Dec 30, 2015 at 12:01 GMT_
_Originally opened as https://github.com/plietar/librespot/issues/18_

----

Currently we don't start loading the next track until the current one has completed. This causes a short delay between tracks.

Instead the next track should start loading before that.

**Comment by sashahilton00** [comment]
<a href="https://github.com/fbrinker"><img src="https://avatars3.githubusercontent.com/u/5345116?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [fbrinker](https://github.com/fbrinker)**
_Saturday Jan 02, 2016 at 10:11 GMT_

----

Maybe with (optional) crossfading, if possible

**Comment by sashahilton00** [comment]
<a href="https://github.com/TonioRoffo"><img src="https://avatars0.githubusercontent.com/u/15906380?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [TonioRoffo](https://github.com/TonioRoffo)**
_Friday Nov 11, 2016 at 19:00 GMT_

----

Crossfading and gapless playback are two different things.  In HW spotify connect gear, gapless is standard.

**Comment by sashahilton00** [comment]
<a href="https://github.com/RafaPolit"><img src="https://avatars1.githubusercontent.com/u/4309019?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [RafaPolit](https://github.com/RafaPolit)**
_Friday Jan 20, 2017 at 21:54 GMT_

----

Gapless is really a must for me, since most of what I listen to is either live or classical music, where much of the 'transitions' are mid-sound (specially in Opera, for instance).

Any time frame on this new feature? Thanks for all the hard work,  best regards,
Rafa.

**Comment by sashahilton00** [comment]
<a href="https://github.com/plietar"><img src="https://avatars0.githubusercontent.com/u/1489775?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [plietar](https://github.com/plietar)**
_Friday Jan 20, 2017 at 22:14 GMT_

----

Gapless isn't too hard, and I'm working on a refactor which will also enable that. I can't give it an ETA, but probably somewhere within the next month.

Cross-fade is a lot trickier, and I don't have any immediate plans for it

**Comment by sashahilton00** [comment]
<a href="https://github.com/cortegedusage"><img src="https://avatars2.githubusercontent.com/u/22986061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [cortegedusage](https://github.com/cortegedusage)**
_Tuesday Feb 14, 2017 at 08:12 GMT_

----

+1 for gapless for me.
(Just for the record, no pressure, keep up the good work.)

**Comment by sashahilton00** [comment]
<a href="https://github.com/johnstok"><img src="https://avatars1.githubusercontent.com/u/459385?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [johnstok](https://github.com/johnstok)**
_Wednesday May 31, 2017 at 18:35 GMT_

----

+1 – would be great if this was implemented.

**Comment by sashahilton00** [comment]
<a href="https://github.com/RafaPolit"><img src="https://avatars1.githubusercontent.com/u/4309019?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [RafaPolit](https://github.com/RafaPolit)**
_Tuesday Jun 27, 2017 at 17:00 GMT_

----

Any news on this? I'm eagerly awaiting for this and 6 months ago you said it wasn't hard.  Was it more difficult than expected or this is simply not a priority for others?  For me its critical.  Please?  Thanks.

**Comment by sashahilton00** [comment]
<a href="https://github.com/sashahilton00"><img src="https://avatars3.githubusercontent.com/u/4185362?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sashahilton00](https://github.com/sashahilton00)**
_Tuesday Jun 27, 2017 at 18:18 GMT_

----

Paul mentioned he was busy until July, most things have been put on hold till then. I'm sure it will be added in due course. Cross fade is harder due to the need to mix audio streams

**Comment by sashahilton00** [comment]
<a href="https://github.com/RafaPolit"><img src="https://avatars1.githubusercontent.com/u/4309019?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [RafaPolit](https://github.com/RafaPolit)**
_Tuesday Jun 27, 2017 at 19:14 GMT_

----

Yeah, I wasn't talking about crossfading, just gapless playback for concerts, and classical pieces that flow seamlessly into the next movement (most critical in Opera).

**Comment by sashahilton00** [comment]
<a href="https://github.com/pauLee"><img src="https://avatars2.githubusercontent.com/u/592889?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [pauLee](https://github.com/pauLee)**
_Friday Nov 17, 2017 at 14:37 GMT_

----

Any progress on gapless playback? I still hear breaks between the tracks.
